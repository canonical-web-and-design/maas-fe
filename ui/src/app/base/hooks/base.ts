import { useCallback, useEffect, useRef, useState } from "react";

import { NotificationSeverity } from "@canonical/react-components";
import type { NotificationProps } from "@canonical/react-components";
import { useDispatch, useSelector } from "react-redux";

import configSelectors from "app/store/config/selectors";
import { actions as messageActions } from "app/store/message";

/**
 * Add a message in response to a state change e.g. when something is created.
 * @param addCondition - Whether the message should be added.
 * @param cleanup - A cleanup action to fire.
 * @param message - The message to be displayed.
 * @param onMessageAdded - A function to call once the message has
                                      been displayed.
 * @param messageType - The notification type.
 */
export const useAddMessage = (
  addCondition: boolean,
  cleanup: () => { type: string },
  message: string,
  onMessageAdded?: (() => void) | null,
  messageSeverity: NotificationProps["severity"] = NotificationSeverity.INFORMATION
): void => {
  const dispatch = useDispatch();

  useEffect(() => {
    if (addCondition) {
      dispatch(messageActions.add(message, messageSeverity));
      onMessageAdded && onMessageAdded();
      dispatch(cleanup());
    }
  }, [
    addCondition,
    cleanup,
    dispatch,
    message,
    messageSeverity,
    onMessageAdded,
  ]);
};

/**
 * Set the browser window title.
 * @param title - The title to set.
 */
export const useWindowTitle = (title?: string): void => {
  const maasName = useSelector(configSelectors.maasName);
  const maasNamePart = maasName ? `${maasName} ` : "";
  const titlePart = title ? `${title} | ` : "";
  useEffect(() => {
    document.title = `${titlePart}${maasNamePart}MAAS`;
  }, [maasNamePart, titlePart]);
};

/**
 * Handle checking when a value has cycled from false to true.
 * @param value - The value to check.
 * @param onCycled - The function to call when the value changes from false to true.
 */
export const useCycled = (
  value: boolean,
  onCycled?: () => void
): [boolean, () => void] => {
  const previousValue = useRef(value);
  const [hasCycled, setHasCycled] = useState(false);
  useEffect(() => {
    if (value && !previousValue.current) {
      onCycled && onCycled();
      setHasCycled(true);
    }
    if (previousValue.current !== value) {
      previousValue.current = value;
    }
  }, [value, onCycled]);
  const resetHasCycled = useCallback(() => {
    setHasCycled(false);
    previousValue.current = false;
  }, [setHasCycled]);
  return [hasCycled, resetHasCycled];
};

/**
 * Handle displaying action progress for batched actions.
 * @param processingCount - The number of items currently being processed.
 * @param onComplete - The function to call when all the items have been processed.
 * @param hasErrors - Whether there are any item errors in state.
 * @param onError - The function to call when an error occurs.
 */
export const useProcessing = (
  processingCount: number,
  onComplete: () => void,
  hasErrors = false,
  onError: () => void
): void => {
  const processingStarted = useRef(false);
  if (processingStarted.current === false && processingCount > 0) {
    processingStarted.current = true;
  }

  // If all the items have finished processing and there are no errors, run the
  // onComplete function.
  useCycled(processingCount === 0, () => {
    if (!hasErrors) {
      onComplete();
    }
  });

  // If the items are processing and errors occur, run the onError function.
  useCycled(hasErrors, () => {
    if (processingStarted) {
      onError();
    }
  });
};

/**
 * Scroll an element into view on render.
 */
export const useScrollOnRender = <T extends HTMLElement>(): ((
  targetNode: T
) => void) => {
  const htmlRef = useRef<HTMLElement>(document.querySelector("html"));
  const onRenderRef = useCallback((targetNode) => {
    if (targetNode && htmlRef?.current) {
      const { height: targetHeight, y: targetTop } =
        targetNode.getBoundingClientRect();
      const windowTop = htmlRef.current.scrollTop;
      const windowBottom = windowTop + window.innerHeight;
      const targetBottom = targetTop + targetHeight;
      // Whether the top of the target is below the bottom of the screen.
      const topOffBottom = targetTop > windowBottom;
      // Whether the top of the target is above the top of the screen.
      const topOffTop = targetTop < windowTop;
      // Whether the top of the target is on the screen.
      const topOnScreen = !topOffTop && !topOffBottom;
      // Whether the bottom of the target is below the bottom of the screen.
      const bottomOffBottom = targetBottom > windowBottom;
      // Whether the target top is on the screen but the bottom is below the bottom.
      const targetPartiallyOffBottom = topOnScreen && bottomOffBottom;
      if (topOffBottom || topOffTop || targetPartiallyOffBottom) {
        window.scrollTo({
          top: targetTop,
          left: 0,
          behavior: "smooth",
        });
      }
    }
  }, []);
  return onRenderRef;
};
