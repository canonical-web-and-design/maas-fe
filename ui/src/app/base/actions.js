export const connectWebSocket = () => {
  return {
    type: "WEBSOCKET_CONNECT"
  };
};

export const fetchAuthUser = () => {
  return {
    type: "WEBSOCKET_SEND",
    payload: {
      actionType: "FETCH_AUTH_USER",
      message: {
        method: "user.auth_user",
        type: 0
      }
    }
  };
};
