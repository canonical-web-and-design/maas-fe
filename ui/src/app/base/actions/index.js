export { default as auth } from "./auth";
export { default as controller } from "./controller";
export { default as device } from "./device";
export { default as dhcpsnippet } from "./dhcpsnippet";
export { default as general } from "./general";
export { default as machine } from "./machine";
export { default as messages } from "./messages";
export { default as packagerepository } from "./packagerepository";
export { default as scripts } from "./scripts";
export { default as subnet } from "./subnet";
export { default as user } from "./user";
export { default as websocket } from "./websocket";
