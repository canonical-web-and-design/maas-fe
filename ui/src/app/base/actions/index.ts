export { default as auth } from "./auth";
export { default as general } from "./general";
export { default as licensekeys } from "./licensekeys";
export { default as messages } from "./messages";
export { default as scripts } from "./scripts";
export { default as status } from "./status";
export { default as websocket } from "./websocket";
