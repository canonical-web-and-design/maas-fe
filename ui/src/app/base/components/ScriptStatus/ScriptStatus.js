import { Tooltip } from "@canonical/react-components";
import PropTypes from "prop-types";

import { ScriptResultStatus } from "app/store/scriptresult/types";

const ScriptStatus = ({
  children,
  hidePassedIcon = false,
  scriptType,
  tooltipPosition,
}) => {
  switch (scriptType.status) {
    case ScriptResultStatus.PASSED: {
      return hidePassedIcon ? (
        children
      ) : (
        <>
          <i className="p-icon--success is-inline" />
          {children}
        </>
      );
    }

    case ScriptResultStatus.DEGRADED:
    case ScriptResultStatus.FAILED:
    case ScriptResultStatus.FAILED_APPLYING_NETCONF:
    case ScriptResultStatus.FAILED_INSTALLING: {
      return (
        <Tooltip
          message="Machine has failed tests."
          position={tooltipPosition}
          positionElementClassName="p-double-row__tooltip-inner"
        >
          <i className="p-icon--error is-inline" />
          {children}
        </Tooltip>
      );
    }

    case ScriptResultStatus.TIMEDOUT: {
      return (
        <Tooltip
          message="Machine has tests that have timed out."
          position={tooltipPosition}
          positionElementClassName="p-double-row__tooltip-inner"
        >
          <i className="p-icon--timed-out is-inline" />
          {children}
        </Tooltip>
      );
    }

    case ScriptResultStatus.APPLYING_NETCONF:
    case ScriptResultStatus.INSTALLING:
    case ScriptResultStatus.RUNNING: {
      return (
        <>
          <i className="p-icon--running is-inline" />
          {children}
        </>
      );
    }

    case ScriptResultStatus.PENDING: {
      return (
        <>
          <i className="p-icon--pending is-inline" />
          {children}
        </>
      );
    }

    default:
      return children;
  }
};

ScriptStatus.propTypes = {
  children: PropTypes.node.isRequired,
  hidePassedIcon: PropTypes.bool,
  scriptType: PropTypes.shape({
    status: PropTypes.number,
    pending: PropTypes.number,
    running: PropTypes.number,
    passed: PropTypes.number,
    failed: PropTypes.number,
  }).isRequired,
  tooltipPosition: PropTypes.oneOf([
    "btm-center",
    "btm-left",
    "btm-right",
    "left",
    "right",
    "top-center",
    "top-left",
    "top-right",
  ]),
};

export default ScriptStatus;
