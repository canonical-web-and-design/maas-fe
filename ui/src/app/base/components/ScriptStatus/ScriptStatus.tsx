import { Icon } from "@canonical/react-components";

import { ScriptResultStatus } from "app/store/scriptresult/types";
import { TestStatusStatus } from "app/store/types/node";

type Props = {
  children?: React.ReactNode;
  status: ScriptResultStatus | TestStatusStatus;
};

const getIconName = (status: ScriptResultStatus | TestStatusStatus): string => {
  switch (status) {
    case ScriptResultStatus.PENDING:
    case TestStatusStatus.PENDING:
      return "pending";
    case ScriptResultStatus.RUNNING:
    case ScriptResultStatus.APPLYING_NETCONF:
    case ScriptResultStatus.INSTALLING:
    case TestStatusStatus.RUNNING:
      return "running";
    case ScriptResultStatus.PASSED:
    case TestStatusStatus.PASSED:
      return "success";
    case ScriptResultStatus.FAILED:
    case ScriptResultStatus.ABORTED:
    case ScriptResultStatus.DEGRADED:
    case ScriptResultStatus.FAILED_APPLYING_NETCONF:
    case ScriptResultStatus.FAILED_INSTALLING:
    case TestStatusStatus.FAILED:
      return "error";
    case ScriptResultStatus.TIMEDOUT:
      return "timed-out";
    case ScriptResultStatus.SKIPPED:
      return "warning";
    case ScriptResultStatus.NONE:
    case TestStatusStatus.NONE:
      return "";
    default:
      return "";
  }
};

const ScriptStatus = ({ children, status }: Props): JSX.Element => {
  const iconName = getIconName(status);

  return (
    <span>
      {iconName && (
        <Icon className={children ? "is-inline" : null} name={iconName} />
      )}
      {children}
    </span>
  );
};

export default ScriptStatus;
