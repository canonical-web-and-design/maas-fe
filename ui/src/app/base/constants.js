/**
 * Message types defined by MAAS websocket API
 */
export const MESSAGE_TYPES = {
  REQUEST: 0,
  RESPONSE: 1,
  NOTIFY: 2,
};

/**
 * Common column sizes used in <Col> components
 */
export const COL_SIZES = {
  CARD_TITLE: 2,
  DELETE_ROW_BUTTONS: 4,
  SIDEBAR: 3,
  TOTAL: 12,
};

export default MESSAGE_TYPES;
