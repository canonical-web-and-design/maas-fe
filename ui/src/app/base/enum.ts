export enum HardwareType {
  Node = 0,
  CPU = 1,
  Memory = 2,
  Storage = 3,
  Network = 4,
  GPU = 5,
}
