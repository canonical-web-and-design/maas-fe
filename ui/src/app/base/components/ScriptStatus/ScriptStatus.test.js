import { shallow } from "enzyme";
import React from "react";

import { scriptStatus } from "app/base/enum";
import ScriptStatus from "./ScriptStatus";

describe("ScriptStatus ", () => {
  it("renders", () => {
    const wrapper = shallow(
      <ScriptStatus scriptType={{ status: scriptStatus.PASSED }}>
        All tests have passed
      </ScriptStatus>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("displays a success icon if scripts passed", () => {
    const wrapper = shallow(
      <ScriptStatus scriptType={{ status: scriptStatus.PASSED }}>
        All tests have passed
      </ScriptStatus>
    );
    expect(wrapper.find(".p-icon--success").exists()).toBe(true);
  });

  it(`does not display a success icon if scripts passed
    and hidePassedIcon is false`, () => {
    const wrapper = shallow(
      <ScriptStatus hidePassedIcon scriptType={{ status: scriptStatus.PASSED }}>
        All tests have passed
      </ScriptStatus>
    );
    expect(wrapper.find(".p-icon--success").exists()).toBe(false);
  });

  it(`does not display an icon if no scripts have run and
    and hideNotRunIcon is true`, () => {
    const wrapper = shallow(
      <ScriptStatus hideNotRunIcon scriptType={{ status: scriptStatus.NONE }}>
        Tests have not run.
      </ScriptStatus>
    );
    expect(wrapper.find(".p-icon--warning").exists()).toBe(false);
  });

  it("displays an error icon and tooltip if scripts have failed", () => {
    const wrapper = shallow(
      <ScriptStatus scriptType={{ status: scriptStatus.FAILED }}>
        Some or all tests have failed
      </ScriptStatus>
    );
    expect(wrapper.find(".p-icon--error").exists()).toBe(true);
    expect(wrapper.find("Tooltip").exists()).toBe(true);
  });

  it("displays a timed out icon and tooltip if scripts have timed out", () => {
    const wrapper = shallow(
      <ScriptStatus scriptType={{ status: scriptStatus.TIMEDOUT }}>
        Some or all tests have timed out
      </ScriptStatus>
    );
    expect(wrapper.find(".p-icon--timed-out").exists()).toBe(true);
    expect(wrapper.find("Tooltip").exists()).toBe(true);
  });

  it("displays a pending icon if scripts are pending", () => {
    const wrapper = shallow(
      <ScriptStatus scriptType={{ status: scriptStatus.PENDING }}>
        Some or all tests are pending
      </ScriptStatus>
    );
    expect(wrapper.find(".p-icon--pending").exists()).toBe(true);
  });

  it("displays a running icon if scripts are running", () => {
    const wrapper = shallow(
      <ScriptStatus scriptType={{ status: scriptStatus.RUNNING }}>
        Some or all tests are running
      </ScriptStatus>
    );
    expect(wrapper.find(".p-icon--running").exists()).toBe(true);
  });

  it("displays a warning icon and tooltip if scripts have not been run", () => {
    const wrapper = shallow(
      <ScriptStatus scriptType={{ status: scriptStatus.NONE }}>
        Some or all tests are running
      </ScriptStatus>
    );
    expect(wrapper.find(".p-icon--warning").exists()).toBe(true);
    expect(wrapper.find("Tooltip").exists()).toBe(true);
  });
});
