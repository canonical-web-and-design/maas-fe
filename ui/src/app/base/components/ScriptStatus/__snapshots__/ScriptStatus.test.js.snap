// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScriptStatus  renders 1`] = `
<Fragment>
  <i
    className="p-icon--success is-inline"
  />
  All tests have passed
</Fragment>
`;
